Statistiche

chenlijun99 opened the pull request.

riccar12 committed:
Minor

* included missing headers
* q creation in p_self
* line 22 p.c

chenlijun99 committed:
build: fix and set up CMake for p and q

riccar12 committed:
Reading stream from popen("find...

Problems saving the matrix

chenlijun99 commented:
@riccar12 

Ho provato a guardare il codice.

Due note sull'uso di `malloc()` e `realloc()`

Fare una chiamata del tipo `malloc(0)` come viene fatto qui

https://github.com/MatMyfta/SistemiOperativi/blob/cb44552ac534a8dece5839cb0acaeeef5686fa7d/stat.c#L19

e qui

https://github.com/MatMyfta/SistemiOperativi/blob/cb44552ac534a8dece5839cb0acaeeef5686fa7d/stat.c#L64

è generalmente sconsigliato.

Infatti, la documentatzione dice:

> If size is zero, the behavior of malloc is implementation-defined. For example, a null pointer may be returned. Alternatively, a non-null pointer may be returned; but such a pointer should not be dereferenced, and should be passed to free to avoid memory leaks.

Vedi https://en.cppreference.com/w/c/memory/malloc

La documentatzione poi non specifica (almeno, io non ho trovato niente) sul comportanmento che `realloc` dovrebbe avere se gli passi un puntatore restituito da un `malloc(0)`. Cioè, dice che se gli viene passato un `NULL`, allora alloca una memoria come farebbe un `malloc`, quindi nel caso `malloc(0)` restituisse `NULL` ci andrebbe bene, ma è tutto "implementation-defined".

Quindi righe come queste:

https://github.com/MatMyfta/SistemiOperativi/blob/cb44552ac534a8dece5839cb0acaeeef5686fa7d/stat.c#L69

potrebbero essere problematiche.

<hr/>

Per quanto riguarda `realloc()`, ho visto i tuoi commenti.

https://github.com/MatMyfta/SistemiOperativi/blob/cb44552ac534a8dece5839cb0acaeeef5686fa7d/stat.c#L36-L42

La documentatzione dice che:

> The reallocation is done by either:
> 
> a) expanding or contracting the existing area pointed to by ptr, if possible.  The contents of the area remain unchanged up to the lesser of the new and old sizes. If the area is expanded, the contents of the new part of the array are undefined.
> b) allocating a new memory block of size new_size bytes, copying memory area with size equal the lesser of the new and the old sizes, and freeing the old block.

Vedi https://en.cppreference.com/w/c/memory/realloc

Quindi, `realloc()` potrebbe benissimo restituire un nuovo blocco di memoria.

chenlijun99 commented:
IMO si poteva adottare un approccio un po' più semplice.
Si alloca una stringa di lunghezza PATH_MAX e poi si usa getline.

Okay, come dice questo [articolo](https://eklitzke.org/path-max-is-tricky), allocare una stringa lunga `PATH_MAX` non è propria la cosa giusta. Però, per affrontare le problematiche presentate nell'articolo l'approccio attuale comunque non è sufficiente e dubito Naimoli pretenda la gestione di questi corner case.

riccar12 commented:
Chiaramente quello che proponi tu è molto più semplice, mapensavo fosse un po restrittivo, nel senso che nel caso di stringhe di dimensione superiore a PATH_MAX vi è un errore, e non sapendo il numero di sottocartelle, anche partendo dall'indirizzo relativo, il path finale potrebbe diventare molto lungo. Ho quindi provato a implementare quel codice per essere il piu flessibili possibile, ma quando ho iniziato non pensavo di trovarmi di fronte a tutte quelle problematiche. Come decidiamo di proseguire quindi, secondo voi?

Edit: in ogni caso una modifica (almeno) del numero del numero di righe deve esserci, altrimenti come puoi istanziare una matrice adeguatamente grande a priori?

chenlijun99 commented:
Il problema è che, se ho letto bene la documentazione, le system calls si aspettano un percorso che non è più lungo di `PATH_MAX`. Quindi anche se poi noi riuscissimo a salvare percorsi che sono più lunghi di `PATH_MAX`, dovremmo comunque passarli a funzioni tipo `open()`, che darebbero comunque errore (ENAMETOOLONG).
Non ho trovato molto su come si potrebbe risolvere questo problema, ma l'idea è fare in modo di cambiare la cartella corrente in modo che si possa passare a `open()` un percorso relativo più corto. Chiaramente, per gestire questo problema per bene, ci vuole ben più di qualche oretta. Considera anche il caso seguente: noi siamo in `/` e abbiamo un percorso ad un file che è tipo `3 * PATH_MAX`. Dovrebbe prima saltare su altre due cartelle intermedie per poi essere su un nodo del filesytstem da cui il percorso relativo al file che vogliamo non sia più lungo di `PATH_MAX`, che quindi può finalmente essere passato a `open()`.
Quindi, una volta che decidiamo di intraprendere la strada di gestire path più lunghi di PATH_MAX, potrebbe essere un bagno di sangue secondo me.
Non penso che Naimoli aspetti questo da noi poveri mortali (se volete gli posso scrivere una e-mail).

riccar12 committed:
Rename stat.c to stat_old.c

riccar12 committed:
Add new stat.c

La matrice salva i path correttamente (a quanto pare). Ho un problema sul passaggio per riferimento della variabile della matrice dal main alla funzione (info al primo commento). Chiedo aiuto, non ne vengo fuori

riccar12 committed:
Update stat.c

riccar12 committed:
Update stat.c

riccar12 commented:
Nel frattempo ho scoperto il motivo per cui non funzionava in docker: il comando find (in docker) non si trova in /bin/find ma in /usr/bin/find, quindi nel momento in cui facciamo il merge di tutto quanto per farlo funzionare in docker, sappiamo che la popen per la lettura dei file contenuti nel path dato in input funziona solamente aggiungendo /usr/ al comando (funziona validate)

riccar12 committed:
Update stat.c

Pietro-Farina committed:
Aggiunta la classe per le statistiche, TODO la parte della codifica ma dipende dal protocollo

riccar12 committed:
Status panel
* first stage
* p status using p_list
* q status through pipe (looking for solutions)
